### ISSUE TITLE
Improve the config property and events binding

### ISSUE THREAD

**Issue by titouanmathis** [studiometa/js-toolkit#56]
### Transform the `config` getter into a static property

It makes more sense as the `config` getter returns the same value for each instance. The `get config()` pattern would be marked as deprecated but would still be available for some time.

```js
// Declaring a config objet
class Component extends Base {
  static config = { 
    name: 'Component' 
  };
}

// Inheriting a parent config object
class ComponentOverride extends Component {
  static config = {
    ...Component.config,
    name: 'ComponentOverride',
  };
}
```

### Introduce the usage of props 

Today, the `config` getter is merged with the `data-options` attribute of the rool element for each instance. This might be complex to maintain for the long run. I suggest to separate the dynamic values of the config into props in order to keep the `config` as readonly and to simplify the management of dynamic values.

A new `props` property would be introduced to the `config` object to define the dynamic values which would be defined with `data-prop-[name]` attributes on the instance's root element. The props would be available via the `this.$props` property.

```html
<div data-component="Component" data-prop-openLabel="Open">
</div>
```

```js
class Component extends Base {
  static config = {
    name: 'Component',
    props: {
      openLabel: String,
      closeLabel: [String, 'Close'], // Default value for the prop if it is not set
      isOpen: Boolean
    },
  };

  mounted() {
    console.log(this.$props.openLabel); // "Open"
    console.log(this.$props.closeLabel); // "Close"
    console.log(this.$props.isOpen); // false
  }
}
```

The following types would be available:

- `String` with `""` as a default value
- `Boolean` with `false` as a default value
- `Array` with `[]` as a default value
- `Object` with `{}`  as a default value

The `Array` and `Object` types will be read with `JSON.parse(value)` and written with `JSON.stringify(value)`.

### Introduce the definition of refs

Refs can be registered today by using a `data-ref="[name]"` attribute. The required refs for a component are not configurable and this can lead to errors at runtime. Defining the required refs for a component would allow for a better developer experience and for a more robust refs management. This would help with the refs resolutions as the component would know the names to look for.

A `refs` property should be added to the static `config` property:

```js
class Component extends Base {
  static config = {
    name: 'Component',
    refs: ['btn', 'input'],
  };
}
```

Element with a `data-ref` attribute with a value not listed in the configuration would still be attached to the instance, but a warning would suggest to either add the ref to the configuration or rename it to match the names in the configuration.

## Advanced event management

Allow the configuration of event handling methods with new suffixes:

- `Passive` to add the `{ passive: true }` configuration to  the `addEventListener` method
- `Once` to add the `{ once: true }` configuration to the `addEventListener` method
- `Capture` to add the `{ capture: true }` configuration to the `addEventListener` method

**Example**
```js
class Component extends Base {
  static config = {
    name: 'Component',
    refs: ['btn'],
  };

  // Only trigger this method once
  onBtnClickOnce() {}

  // Combine the 3 suffixes in any order
  onBtnClickOncePassiveCapture() {}
  onBtnClickPassiveOnceCapture() {}
}
```

**Comment by perruche** [comment]
Not a fan of the props suggestion:
- This will require a lot of change to update projects on previous version of the js-toolkit, this is prety much breaking change 
- I feel like this is redoing Vue. If we need dynamic props should we use Vue ? cf #55 

Refs and events suggestions are great.

**Comment by titouanmathis** [comment]
> - This will require a lot of change to update projects on previous version of the js-toolkit, this is prety much breaking change

My proposal is indeed a breaking change, but it should not be, my bad. The idea behind the prop definition was to replace the `data-options` attribute and the `this.$options` property. The main goal were : 

- Pre-define the allowed options and set their default value
- Allow for customization on a per option base with a specific `data-option-*` attribute (i.e. `data-option-label="Hello")
- Separate the class configuration from the instance options

So instead of talking about `props`, I should have stick to the `options` name to avoid any misunderstanding.

> - I feel like this is redoing Vue. If we need dynamic props should we use Vue ? cf #55 

It is, as said above, only a way to declare options with a more robust definition and not the same as the Vue props binding and reactivity.

---

I will update the proposal to remove any breaking changes and keep using the `options` logic. Thanks for your input!

**Comment by perruche** [comment]
I agree that config ≠ options is a bit confusing atm 😄

**Comment by titouanmathis** [comment]
As announced above, here is a rewrite on the prop/option proposal.

---

### Improve the options management

#### Motivation

Today, the `config` getter is merged with the `data-options` attribute of the rool element for each instance and the result is accessible via the `this.$options` property. This behavior can be confusing as the config and options have different roles:

- the `config` getter defines how the component should work inside of a `Base` ecosystem
- the `data-options` attribute defines instance specific options to be used inside the component

Furthermore, defining options in JSON format in a `data-options` attribute can be complex / cumbersome and the available options for a component are not defined.

#### Design

The first step would be to implement the definition of available options in an `options` property of the `config` getter/static property and deprecate the old method:

```js
// Before
class Component extends Base {
  get config() {
    return {
      name: 'Component',
      optionOne: false,
      // <name>: <default>,
    };
  }
}

// After
class Component extends Base {
  static config = {
    name: 'Component',
    options: {
      optionOne: [Boolean, false],
      // <name>: [<type>, <default>],
     // <name>: <type>,
    },
  };
}
```

This can be done without breaking change as the options are not read directly from the `config` getter but via the `$options` getter. We can update its internal functioning to adapt to the new format, while displaying a warning when using the old format.

The second step would then be to add the possibility to override an option with a `data-option-<name>="<value>"` attribute and to deprecate the definition of all options via the `data-options` attribute.

```html
<!-- Before -->
<div data-component="Component" data-options='{ "label": "Close" }'></div>

<!-- After -->
<div data-component="Component" data-option-label="Close"></div>
```

This is where defining the type for options would be useful as numbers, booleans, objects and arrays could be parsed correctly and not just returned as strings. The following types would be available:

- `String` with `""` as a default value
- `Number` with `0` as a default value 
- `Boolean` with `false` as a default value
- `Array` with `[]` as a default value
- `Object` with `{}`  as a default value

The `Array` and `Object` types would be read with `JSON.parse(value)` and written with `JSON.stringify(value)`. 

Usage for each type would be as follow:

```html
<div data-component="Component"
  data-option-string="Some text"
  data-option-number="10.5"
  data-option-array="[0, 1, 2]"
  data-option-boolean
  data-option-object="{&quot;foo&quot;:&quot;foo&quot;,&quot;bar&quot;:true}">
</div>
```

```js
class Component extends Base {
  static config = {
    name: 'Component',
    options: {
      string: String, // Default to an empty string if `data-option-string` not specified
      number: [Number, 1], // Default number is 1
      array: Array,
      boolean: Boolean,
      otherBoolean: Boolean,
      object: Object,
    },
  };

  mounted() {
    console.log(this.$options.string) // "Some text"
    console.log(this.$options.number) // 10.5
    console.log(this.$options.array) // [0, 1, 2]
    console.log(this.$options.boolean) // true
    console.log(this.$options.otherBoolean) // false 
    console.log(this.$options.object) // { foo: 'foo', bar: true }
  }
}
```

The `log` and `debug` properties could be set via the `data-log` and `data-debug` properties.

To avoid any breaking changes, the `this.$options` property should be created as follow : 

- include properties from the `config` getter not in the list `['components', 'refs', 'options']`, throw a deprecation warning if some are found
- include the `config.name` property but throw a deprecation warning when accessing it as it should be read from `Component.config.name`
- include properties from the `data-options` attribute, throw a deprecation warning if some are found
- set the `$options.log` and `$options.debug`  based on the `data-log` and `data-debug` attributes
- define getters for the properties defined in `config.options` that read the value from the `data-option-*` attributes
- define setters for the properties defined in `config.options` to update the `data-option-*` attributes

**Comment by titouanmathis** [comment]
Available in [v1.0.0](https://github.com/studiometa/js-toolkit/releases/tag/1.0.0).
